File: Scala/src/main/scala/TextbookTrack/Chapter11/BA11A.scala
package TextbookTrack.Chapter11

object BA11A {
  import TextbookTrack.Utils.IntegerMassTable

  final case class Edge(nodeFrom: Int, nodeTo: Int, label: Char)

  class Graph(initialSpectrum: List[Int]) {
    import Graph.{buildSpectrumGraph, verifySpectrum}

    val spectrum: List[Int] = verifySpectrum(initialSpectrum)
    val adjacencyList: Map[Int, List[Edge]] = buildSpectrumGraph(spectrum)

    def edges: Iterator[String] =
      for {
        (node, neighbours) <- adjacencyList.iterator
        Edge(_, neighbour, label) <- neighbours
      } yield s"$node->$neighbour:$label"
  }

  object Graph {
    private def verifySpectrum(spectrum: List[Int]): List[Int] = {
      val zeroMass: Int = 0
      val updatedSpectrum: List[Int] = if (spectrum.contains(zeroMass)) spectrum else zeroMass :: spectrum
      updatedSpectrum.sorted
    }

    private def buildSpectrumGraph(spectrum: List[Int]): Map[Int, List[Edge]] = {
      val inverseMapping: Map[Int, Char] = reverseMassTableMapping(IntegerMassTable)
      val edges: List[Edge] = for {
        (node1, ix) <- spectrum.zipWithIndex
        node2 <- spectrum.drop(ix)
        weightDifference: Int = node2 - node1
        label <- inverseMapping.get(weightDifference)
      } yield Edge(node1, node2, label)
      edges.groupBy(_.nodeFrom)
    }
  }

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def reverseMassTableMapping(massTable: Map[Char, Int]): Map[Int, Char] = massTable.map(_.swap)

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val spectrum: List[Int] = convertToIntList(reader.next())
    val graph = new Graph(spectrum)
    graph.edges.foreach(println)
  }
}

File: Python/textbook_track/chapter11/ba11a.py
# Construct the Graph of a Spectrum
import sys

from collections import defaultdict

from textbook_track.resources.utils import read_amino_acid_mass_table


class Graph:
    def __init__(self, spectrum, mass_table):
        self._spectrum = self._verify_spectrum(spectrum)
        self._adjacency_list = self._build_spectrum_graph(mass_table)

    def edges(self):
        for node, neighbours in self._adjacency_list.items():
            for neighbour, label in neighbours:
                yield f'{node}->{neighbour}:{label}'

    @staticmethod
    def _verify_spectrum(spectrum):
        if (zero_mass := 0) not in spectrum:
            spectrum.append(zero_mass)
        return sorted(spectrum)

    def _build_spectrum_graph(self, mass_table):
        inverse_mapping = reverse_mass_table_mapping(mass_table)
        adjacency_list = defaultdict(list)
        for ix, node1 in enumerate(self._spectrum):
            for node2 in self._spectrum[ix+1:]:
                weight_difference = node2 - node1
                if (label := inverse_mapping.get(weight_difference)) is not None:
                    adjacency_list[node1].append((node2, label))
        return dict(adjacency_list)


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def reverse_mass_table_mapping(mass_table):
    return {mass: amino_acid for amino_acid, mass in mass_table.items()}


def main():
    reader = sys.stdin
    spectrum = convert_to_intlist(next(reader))
    mass_table = read_amino_acid_mass_table()
    graph = Graph(spectrum, mass_table)
    for edge in graph.edges():
        print(edge)


if __name__ == '__main__':
    main()
